### DISCUSSION TITLE
Backslash sequences

### DISCUSSION THREAD

**Discussion by SicroAtGit** [SicroAtGit/RegEx-Engine#5] [Ideas]
In addition to character classes, there will also be shorthand character classes. However, I'm not quite sure yet which ones there should be and which characters they should cover.

According to this website, the different RegEx engines cover different characters in the shorthand character classes:
https://www.regular-expressions.info/shorthand.html

The current listing:

- `\r` for carriage return (#8)
- `\n` for new line (#9)
- `\t` for horizontal tab character (#10)
- `\f` for form feed (#22)
- `\d` for digit (#14)
- `\D` for no digit (#23)
- `\s` for whitespace character (#15)
- `\S` for no whitespace character (#24)
- `\w` for word character (#25)
- `\W` for no word character (#26)
- `\xhh` (#17)
- `\uhhhh` (#18)
- `\Q`...`\E` (#13)

**Comment by tajmone** [comment]
> According to this website, the different RegEx engines cover different characters in the shorthand character classes:
https://www.regular-expressions.info/shorthand.html

BTW, I've bought both [RegexBuddy] and [RegexMagic] form JGS (author of the website you linked), so if you need me to test some RegExs for you I'll happily do it. Both tools have a custom engine that includes [_all_ versions of the major RegEx engines] (so that you can test backward compatibility issues with any engine) plus the [custom engine by JGS], which is very powerful (also documented at the website).

One of these two programs also allows debugging a RegEx to break it down into each single passage, in case you need to compare expected behaviour in your code with actual behaviour by other engines.

As for the shorthand classes to implement, it really depends on what your engine goals are — which I'm guessing is mostly oriented toward lexers creation?

I'm not quite sure that `\h` and `\v` would be all that useful (vertical tabs are not used much in Western languages, and `\t` should suffice in place of `\h`), also these tend to have different meanings across engines.

Some other useful shorthands can be found here:

- [Special Characters] — `\Q`...`\E` for escaping literal sequences.
- [Non-Printable Characters] — `\x` ASCII chars, `\u` Unicode.
- [Anchors] — `\A` and `\Z` are very useful for lexing.
- [Word Boundaries] — Tcl style `\m` and `\M`.

I know that the above don't all qualify as characters shorthand, for some of them are more abstract in nature, but still...




[RegexBuddy]: https://www.regexbuddy.com
[RegexMagic]: https://www.regexmagic.com/
[_all_ versions of the major RegEx engines]: https://www.regexmagic.com/flavors.html
[custom engine by JGS]: https://www.regular-expressions.info/jgsoft.html

[Special Characters]: https://www.regular-expressions.info/characters.html
[Non-Printable Characters]: https://www.regular-expressions.info/nonprint.html
[Anchors]: https://www.regular-expressions.info/anchors.html
[Word Boundaries]: https://www.regular-expressions.info/wordboundaries.html

**Reply by SicroAtGit** [reply]
> [RegexBuddy](https://www.regexbuddy.com) and [RegexMagic](https://www.regexmagic.com/)

Yes, very cool programs. If I had a lot to do with complex regexes, I would definitely buy the programs as well.

> so if you need me to test some RegExs for you I'll happily do it

Thank you for the offer.

> As for the shorthand classes to implement, it really depends on what your engine goals are — which I'm guessing is mostly oriented toward lexers creation?

Yes, the focus of the regex engine is to create lexers.

> Some other useful shorthands can be found here:

Thank you for these suggestions. However, these are off-topic here and should be discussed in other discussions.

**Reply by tajmone** [reply]
>  However, these are off-topic here and should be discussed in other discussions.

You're right, but since they are similar in notation (and probably also implementation-wise) I though of mentioning them here. 

Also, when it comes to RegExs, the terminology to describe syntax elements can vary significantly from one context to another (e.g. an academic treatise on the topic might use terms from Sets theory and mathematics, whereas engine implementations will use terms more related to practical coding, etc.).  E.g. the `*` metacharacter is called _Kleen star_ in some contexts, and many different names in other contexts, varying from _star operator_, to many others.

So it's hard to determine what goes where sometimes, because grouping might also depend from the angle you're looking at them (implementation, functionality, theoretical category, etc.).

For example, special characters like `\t`, `\r` and `\n` aren't really shorthand classes, they are just escape sequences, for they don't repesent sets, unlike `\d` for example. Hence I think it might be convenient to treat all of these in a same discussion, at least in these preliminary stages, yet group them according to category.

**Reply by SicroAtGit** [reply]
> For example, special characters like `\t`, `\r` and `\n` aren't really shorthand classes, they are just escape sequences, for they don't repesent sets, unlike `\d` for example.

Yes, in most cases they are called escape sequences. But they can also be seen as character classes that contain only one character.

The dot can also be seen as a character class, instead of a meta-symbol.

I have changed the title of the discussion from "Shorthand character classes" to "Backslash sequences". This allows the discussion to continue unchanged.

**Reply by tajmone** [reply]
> The dot can also be seen as a character class, instead of a meta-symbol.

Yes, that's exactly the problem I mentioned earlier, about different contexts using different terminology, especially in books and articles on this topic. For the sake of this project, it would be a good idea if you established your own choice of terminology and stick with it in both source code and documentation, possibly providing a glossary in the Wiki. Of course, you can always mention the alternative way each element is called, but it should be clear which is the official term adopted in this project.

RegExs are already a rather complex topic by nature, and surely all the different terms used to describe things are not making it easier for newbies to the topic — I remember that when I was studying RegExs, much effort was wasted trying to figure out what all these different terms meant.

**Reply by SicroAtGit** [reply]
> I'm not quite sure that `\h` and `\v` would be all that useful (vertical tabs are not used much in Western languages, and `\t` should suffice in place of `\h`), also these tend to have different meanings across engines.

Okay, I remove "h" and "v" from the list.

> * Special Characters — `\Q`...`\E` for escaping literal sequences.

Maybe a feature for version 1.1.
I have to see how useful it is.

> * Non-Printable Characters — `\x` ASCII chars, `\u` Unicode.

In the engine I work with characters instead of bytes, so I cannot separate ASCII and Unicode characters.

I add `\xHEXCODE` to the list.

> * Anchors — `\A` and `\Z` are very useful for lexing.

To do this, the engine has to look around, which is not possible with DFA. All matching characters will also be in the result.

There is a special variant of DFA, called Tagged DFA (re2c uses this), with which this is possible, but I don't want to make it too complicated and therefore stick to a normal DFA as far as possible.

> * Word Boundaries — Tcl style `\m` and `\M`.

I needed them to force the longest match in my old lexer, no matter how they are ordered in the regex. But because this engine always returns the longest match, they are no longer needed in this case. And if no keyword matches, it is usually the regex for identifiers that matches then. And if no identifiers are allowed at this point, it is the parser's job to output a syntax error.

Also, as above, the engine has to look around for this.

> For the sake of this project, it would be a good idea if you established your own choice of terminology and stick with it in both source code and documentation, possibly providing a glossary in the Wiki. Of course, you can always mention the alternative way each element is called, but it should be clear which is the official term adopted in this project.

Yes, good idea. I'm already thinking about how to name all the things while I'm writing locally on my PC the texts of the issues/discussions I'll publish soon that describe the planned features.

**Reply by tajmone** [reply]
> > Non-Printable Characters — `\x` ASCII chars, `\u` Unicode.
>
> In the engine I work with characters instead of bytes, so I cannot separate ASCII and Unicode characters.
>
> I add `\xHEXCODE` to the list.

I'm not sure I follow you here. Both are meant as shorthands to type non-printable characters (or _glyphs_, if we want to avoid confusion) or characters which are not available in the user keyboard, or which are not covered by most editors' fonts.

So, the idea of the `\x` shorthand is to handle ASCII chars, whereas `\u` Unicode points (which has fixed number of hex digits which is much longer). In any case, it will be up to you to translate these into PB chars, also keeping in mind that PB uses UCS-2, so it doesn't fully support Unicode (not all planes).

Usually the `\x` shorthand is more practical for those well known non-printable chars which are representable as single-bytes (e.g. non-breaking spaces, etc.), for which we usually know by memory their hex code, but not their Unicode point.

You might as well use `\x` to cover ISO-8859-1/Latin-1 chars, to avoid the typical confusion in PB about ASCII meaning "extended ASCII" (which doesn't exists), and assume that any char beyond 127 is a Latin-1 char.

I noticed that in the sources you often refer to `SizeOf(Char)`, which rather surprised me since it seems to intend that it could also be compiled in ASCII mode — which, apart from not being supported any more, would also make this code unlikely to compile on those old versions of PB.

Wouldn't it just make more sense to assume that every char is two bytes? This would also allow you some speed optimizations, since you can avoid multiplying or dividing by 2 (or multiples), using bit-shift operations instead (which are faster at the processor level).

**Reply by SicroAtGit** [reply]
My engine sees characters as PB sees characters and thus has the same limitations (no full Unicode support).

If you use the engine in PB in ASCII mode you are also limited to ASCII characters in the RegEx.

`SizeOf(Character)` evaluates to `1` in ASCII mode and to `2` in Unicode mode. So my engine always reads one byte for one character in ASCII mode and two bytes for one character in Unicode mode.

Even though ASCII mode is no longer supported in newer PB versions, I still write codes that are compatible with it.

In my engine `\x00FF` would be identical to the PB code `Chr($00FF)`. Therefore, if you want to use the engine with PB in ASCII mode, you can't write `\xFFFF` in the regex then, but only `\x00FF` at most.

We can also take `\#` instead of `\x` to avoid confusion.

**Reply by tajmone** [reply]
> Even though ASCII mode is no longer supported in newer PB versions, I still write codes that are compatible with it.

That sounds a lot of work to preserve backward compatibility, especially with v5 being imminent. Probably with PB5 you won't be able to ensure backward compatibility with the old PB4 ASCII mode, except through a lot of compiler directives. Is it worth it?

Also, I think that the performance boost of bit shift operations is well worth dropping ASCII support, since the code that does the `SizeOf(Character)` multiplication recurs quite often. Usually, good compilers can optimize multiplications and divisions by multiples of twos to bit shifts, but the PB compiler doesn't handle optimizations. See this article:

https://softwareengineering.stackexchange.com/questions/234967/speeds-of-multiplication-and-division

Maybe PB5 will benefit from GCC optimizations in this case, but that won't work with the Fasm backend. In any case, in this type of pointers arithmetic, involving offsets, it's customary to find bit shifts.

You could define a macro at the beginning of the module, to decide at once whether Char is 1 or two bytes, and then use the macro to either do nothing (in case of ASCII, 1 byte) or do a bit shift operation. This would also make the ASCII code faster, since a multiplication by one is a waste of clock cycles (in a compiler that doesn't optimize).

**Reply by SicroAtGit** [reply]
> That sounds a lot of work to preserve backward compatibility

No, just using `*pointer + SizeOf(Character)` instead of `*pointer + 2`.

> Probably with PB5 you won't be able to ensure backward compatibility with the old PB4 ASCII mode, except through a lot of compiler directives.

You mean PB 6 (upcoming version) and PB 5.46 (last version with ASCII support).

In PB 6 UCS-2 is also used.

But I understand you now. I don't mean that my code in the new PB versions continue to offer ASCII support, but that my code with old PB versions in ASCII mode also work.

**Regarding `SizeOf(Character)` and multiplication/division**

`SizeOf()` is a compiler function whose value is evaluated at compile time. With this value I only perform addition calculations in the code (to move the pointer to the next character of the string), except when displaying error messages I perform one division calculation.

Where do you see a lot of multiplication calculations in the code? Maybe the used prefix `*` for pointer variables confused you?

**Reply by tajmone** [reply]
> You mean PB 6 (upcoming version) and PB 5.46

Yes, I must I've typed the wrong keys (my eyesight has degraded a lot in the last year, so I can't really read well on the screen).

> Where do you see a lot of multiplication calculations in the code? 

In offsets, where the number of chars is multiplied by `SizeOfChar()`.

**Reply by SicroAtGit** [reply]
> In offsets, where the number of chars is multiplied by `SizeOfChar()`.

https://github.com/SicroAtGit/PB-RegEx-Engine/blob/b84cefe688ec9978bb0cac644f5497b26e97d91a/RegExEngine.pbi#L242

Where do you see multiplication in this line of code?

`currentPosition` is a memory pointer pointing to a string. And with `SizeOf(Character)` I add `1` (ASCII mode) or `2` (Unicode mode) to the value of `currentPosition` so that the memory pointer `currentPosition` points to the next character in the string afterwards.

It works like this little example code:
```purebasic
*text.Character = @"Example text"
Debug Chr(*text\c) ; Outputs "E"
*text + SizeOf(Character)
Debug Chr(*text\c) ; Outputs "x"
```

**Reply by SicroAtGit** [reply]
I was confused by the article on the Regular Expression website. I think I have understood it now. `\x` saves writing the leading `00` at `\u` if you only want to define character codes between `00` and `FF`. So `\x21` is the same as `\u0021`.

**Reply by tajmone** [reply]
Yes, but then it's up to you to decide how to implement these. My suggestion would be to have `\x` handle ISO-8859-1 characters, and `\u` Unicode points. If you don't specify a single char encoding for `\x`, there's no way to know what any value above 127 would translate to, in terms of Unicode chars. Bear in mind that many of these RegEx engines do date back to the era of single-character encodings, code pages, etc., so their conventions might be rooted in some old standards.

**Reply by SicroAtGit** [reply]
OK, thank you very much!

I have now added everything you suggested in your comment that is implementable with a standard DFA to the list.

Please create a new comment (no reply) for more suggestions.

**Comment by SicroAtGit** [comment]
I have now looked again at [the documentation of Rust's crate `regex`](https://docs.rs/regex/latest/regex/#perl-character-classes-unicode-friendly) and now have a picture of which characters (Unicode's character classes) the various predefined RegEx character classes should have.

With the help of the very useful tool [`UnicodeSet`](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp) of the official Unicode website I can easily have the Unicode's character classes converted to RegEx character classes. For this output I wrote a tool which generates the tables for the character classes.

I have adjusted the listing at the top and the relevant issues to use the Unicode's character classes. For example, `\d` now matches not only `[0-9]`, but all characters of the Unicode character class `Nd`, except those exceeding `\uFFFF`. The RegEx engine will still only support Unicode characters up to `\uFFFF`.

**Comment by SicroAtGit** [comment]
A backslash sequence that matches a byte value can also be useful. So the RegEx engine can also be used to match byte sequences (beside characters) in binary data. This is not so unusual anymore. The question comes up again and again on the Internet (search for: _regex binary data_).

The NFA/DFA of the RegEx engine already works byte-based, so the implementation should not be difficult.

**Example:**

Syntax: `\b00` up to `\bFF`
1. Match the byte sequence: `3C`, `3D`
2. Match the string: `Test`
```
\b3C\b3DTest
```
`\b` has been used only for the example. We should use another letter to avoid confusion because it is a common backslash sequence for word boundary.

**Reply by tajmone** [reply]
> The NFA/DFA of the RegEx engine already works byte-based, so the implementation should not be difficult.

Interesting. I haven't yet had the time to look at the implementation details, but I assumed that since PB uses the UCS2 encoding internally to represent strings and "Unicode" characters every character would map to two bytes — i.e. if it's an ASCII char it would be followed by a null char (e.g. `A` &rarr; 0x41 + 0x00).

When using RegExs to parse binary data, end users probably want a per-byte fine-grain control over matched data, i.e. no null-padding to ensure a Char is a two-bytes UCS entity (`A` &rarr; 0x41). Isn't this going to create alignment problems?

**Reply by SicroAtGit** [reply]
> I assumed that since PB uses the UCS2 encoding internally to represent strings and "Unicode" characters every character would map to two bytes — i.e. if it's an ASCII char it would be followed by a null char (e.g. `A` → 0x41 + 0x00).

Yes, that's how the engine works:

- `A` → 0x41 + 0x00
- `\x41` → 0x41 + 0x00
- `\u0041` → 0x41 + 0x00

> When using RegExs to parse binary data, end users probably want a per-byte fine-grain control over matched data, i.e. no null-padding to ensure a Char is a two-bytes UCS entity (`A` → 0x41). Isn't this going to create alignment problems?

This is what the `\b` from the example above is for:

- It matches a _single_ byte, instead of a character (2 bytes)
- `\b41` → 0x41
- `\b3C\b3DTest` → 0x3C + 0x3D + 0x54 + 0x00 + 0x65 + 0x00 + 0x73 + 0x00 + 0x74 + 0x00

---

Alternatively, we could implement a new RegEx mode that only allows characters from `\x00` to `\xFF` and where every character matches only a single byte:

- `(?b)Test` → 0x54 + 0x65 + 0x73 + 0x74
- `(?b)\x41\x00` → 0x41 + 0x00
